### PR TITLE
[WIP]share code among toolkits for spring cloud

### PR DESCRIPTION
**Pull request by wangmingliang-ms**

### CODE REVIEW

File: azure-toolkit-libs/azure-toolkit-springcloud-lib/src/main/java/com/microsoft/azure/toolkit/lib/springcloud/service/SpringCloudDeploymentService.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+    public void restart(final SpringCloudDeploymentEntity deployment) {
+        this.restart(deployment.getName(), deployment.getApp());
+    }
+
+    public void remove(final SpringCloudDeploymentEntity deployment) {
+        this.remove(deployment.getName(), deployment.getApp());
+    }
+
+    public void start(String deployment, final SpringCloudAppEntity app) {
```
Comment (andxu) [inline azure-toolkit-libs/azure-toolkit-springcloud-lib/src/main/java/com/microsoft/azure/toolkit/lib/springcloud/service/SpringCloudDeploymentService.java:38]: why it is public? if it need to be public, SpringCloudAppEntity should be taken as the first argument.

File: azure-toolkit-libs/azure-toolkit-springcloud-lib/src/main/java/com/microsoft/azure/toolkit/lib/springcloud/service/SpringCloudClusterService.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+public class SpringCloudClusterService {
+
+    private final AppPlatformManager client;
+
+    public SpringCloudClusterService(AppPlatformManager client) {
+        this.client = client;
+    }
+
+    public void remove(final SpringCloudClusterEntity service) {
```
Comment (andxu) [inline azure-toolkit-libs/azure-toolkit-springcloud-lib/src/main/java/com/microsoft/azure/toolkit/lib/springcloud/service/SpringCloudClusterService.java:31]: remove -> delete?

File: azure-toolkit-libs/azure-toolkit-springcloud-lib/src/main/java/com/microsoft/azure/toolkit/lib/springcloud/service/SpringCloudClusterService.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        this.client.services().deleteAsync(
+            service.getResourceGroup(),
+            service.getName()).await();
+    }
+
+    public SpringCloudClusterEntity reload(final SpringCloudClusterEntity service) {
```
Comment (andxu) [inline azure-toolkit-libs/azure-toolkit-springcloud-lib/src/main/java/com/microsoft/azure/toolkit/lib/springcloud/service/SpringCloudClusterService.java:37]: where's the getSpringCloudCluster method? I need to create SpringCloudClusterEntity for a get purpose?

File: azure-toolkit-libs/azure-toolkit-springcloud-lib/src/main/java/com/microsoft/azure/toolkit/lib/springcloud/service/SpringCloudClusterService.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+    public List<SpringCloudAppEntity> getApps(final SpringCloudClusterEntity service) {
+        final PagedList<AppResourceInner> page = this.client.apps().inner().list(
+            service.getResourceGroup(),
+            service.getName());
+        page.loadAll();
+        return page.stream().map(SpringCloudAppEntity::fromResource).collect(Collectors.toList());
+    }
+
+    public List<SpringCloudClusterEntity> getAllClusters() {
```
Comment (andxu) [inline azure-toolkit-libs/azure-toolkit-springcloud-lib/src/main/java/com/microsoft/azure/toolkit/lib/springcloud/service/SpringCloudClusterService.java:52]: listSpringCloudClusters

File: azure-toolkit-libs/azure-toolkit-springcloud-lib/src/main/java/com/microsoft/azure/toolkit/lib/springcloud/service/SpringCloudAppService.java [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+    public void remove(final SpringCloudAppEntity app) {
+        this.client.apps().deleteAsync(
+            app.getCluster().getResourceGroup(),
+            app.getCluster().getName(),
+            app.getName()
+        ).await();
+    }
+
+    public SpringCloudAppEntity get(final SpringCloudAppEntity app) {
```
Comment (andxu) [inline azure-toolkit-libs/azure-toolkit-springcloud-lib/src/main/java/com/microsoft/azure/toolkit/lib/springcloud/service/SpringCloudAppService.java:49]: a cluster and a app name as parameter?

File: azure-toolkit-libs/azure-toolkit-springcloud-lib/src/main/java/com/microsoft/azure/toolkit/lib/springcloud/model/SpringCloudAppEntity.java [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+        return inner;
+    }
+
+    public SpringCloudAppEntity inner(AppResourceInner inner) {
+        this.inner = inner;
+        return this;
+    }
+
+    public boolean isPublic() {
+        return this.inner.properties().publicProperty();
```
Comment (andxu) [inline azure-toolkit-libs/azure-toolkit-springcloud-lib/src/main/java/com/microsoft/azure/toolkit/lib/springcloud/model/SpringCloudAppEntity.java:52]: how to handle NPE?

File: azure-toolkit-libs/azure-toolkit-springcloud-lib/src/main/java/com/microsoft/azure/toolkit/lib/springcloud/model/JavaRuntimeVersion.java [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.microsoft.rest.ExpandableStringEnum;
+
+import java.util.Collection;
+
+/**
+ * Defines values for JavaRuntimeVersion.
+ */
+public final class JavaRuntimeVersion extends ExpandableStringEnum<JavaRuntimeVersion> {
```
Comment (andxu) [inline azure-toolkit-libs/azure-toolkit-springcloud-lib/src/main/java/com/microsoft/azure/toolkit/lib/springcloud/model/JavaRuntimeVersion.java:18]: this JavaRuntimeVersion  class should be shared among webapp/functions

File: azure-spring-cloud-maven-plugin/src/main/java/com/microsoft/azure/maven/springcloud/ConfigMojo.java [hunk 8]
```diff
@@ -310,24 +311,21 @@ private void configureAppName() throws IOException, InvalidConfigurationExceptio
...
         }
-
-        final ServiceResourceInner targetAppCluster = this.wrapper.handleSelectOne("select-ASC", clusters, null, ServiceResourceInner::name);
+        final List<SpringCloudCluster> clusters = az.clusters();
+        this.wrapper.putCommonVariable("clusters", clusters);
+        final SpringCloudCluster targetAppCluster = this.wrapper.handleSelectOne("select-ASC", clusters, null, c -> c.entity().getName());
         if (targetAppCluster != null) {
-            this.appSettings.setClusterName(targetAppCluster.name());
-            getLog().info(String.format("Using service: %s", TextUtils.blue(targetAppCluster.name())));
+            this.appSettings.setClusterName(targetAppCluster.entity().getName());
```
Comment (andxu) [inline azure-spring-cloud-maven-plugin/src/main/java/com/microsoft/azure/maven/springcloud/ConfigMojo.java:327]: we'd better not use the term 'entity', use basic?

File: azure-toolkit-libs/azure-toolkit-springcloud-lib/src/main/java/com/microsoft/azure/toolkit/lib/common/IEntity.java [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+ * THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO
+ * THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+ * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package com.microsoft.azure.toolkit.lib.common;
+
+public interface IEntity {
```
Comment (andxu) [inline azure-toolkit-libs/azure-toolkit-springcloud-lib/src/main/java/com/microsoft/azure/toolkit/lib/common/IEntity.java:25]: IAzureEntity?

File: azure-toolkit-libs/azure-toolkit-springcloud-lib/src/main/java/com/microsoft/azure/toolkit/lib/springcloud/SpringCloud.java [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+
+    private SpringCloud(AppPlatformManager client) {
+        this.client = client;
+    }
+
+    public static SpringCloud az(AppPlatformManager client) {
+        return new SpringCloud(client);
+    }
+
+    public SpringCloudCluster cluster(SpringCloudClusterEntity cluster) {
```
Comment (andxu) [inline azure-toolkit-libs/azure-toolkit-springcloud-lib/src/main/java/com/microsoft/azure/toolkit/lib/springcloud/SpringCloud.java:43]: we don't need this kind of method? or maybe it should be private

File: azure-toolkit-libs/azure-toolkit-springcloud-lib/src/main/java/com/microsoft/azure/toolkit/lib/springcloud/ConfigurationUtils.java [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+
+    private static final int DEFAULT_PERSISTENT_DISK_SIZE = 50;
+    private static final int DEFAULT_SKU_CAPACITY = 1;
+    private static final int DEFAULT_MEMORY_IN_GB = 1;
+
+    private static final JavaRuntimeVersion DEFAULT_RUNTIME_VERSION = JavaRuntimeVersion.JAVA_8;
+
+    private static final String RUNTIME_VERSION_PATTERN = "[Jj]ava((\\s)?|_)(8|11)$";
+    private static final int TIMEOUT_SCALING = 60; // Use same timeout as service
+    protected static final List<DeploymentResourceStatus> DEPLOYMENT_PROCESSING_STATUS =
```
Comment (andxu) [inline azure-toolkit-libs/azure-toolkit-springcloud-lib/src/main/java/com/microsoft/azure/toolkit/lib/springcloud/ConfigurationUtils.java:64]: private?

File: azure-toolkit-libs/azure-toolkit-springcloud-lib/src/main/java/com/microsoft/azure/toolkit/lib/springcloud/ConfigurationUtils.java [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+import org.apache.commons.lang3.StringUtils;
+
+import javax.annotation.Nonnull;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class ConfigurationUtils {
```
Comment (andxu) [inline azure-toolkit-libs/azure-toolkit-springcloud-lib/src/main/java/com/microsoft/azure/toolkit/lib/springcloud/ConfigurationUtils.java:49]: AzureSpringCloudUtils

File: azure-toolkit-libs/azure-toolkit-springcloud-lib/src/main/java/com/microsoft/azure/toolkit/lib/springcloud/ConfigurationUtils.java [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+            return false;
+        }
+        final boolean isInstanceDeployed = instanceList.stream().noneMatch(instance ->
+            StringUtils.equalsIgnoreCase(instance.status(), "waiting") || StringUtils.equalsIgnoreCase(instance.status(), "pending"));
+        final boolean isInstanceDiscovered = instanceList.stream().allMatch(instance ->
+            StringUtils.equalsIgnoreCase(instance.discoveryStatus(), finalDiscoverStatus));
+        return isInstanceDeployed && isInstanceDiscovered;
+    }
+
+    public static JavaRuntimeVersion getRuntimeVersion(String runtimeVersion) {
```
Comment (andxu) [inline azure-toolkit-libs/azure-toolkit-springcloud-lib/src/main/java/com/microsoft/azure/toolkit/lib/springcloud/ConfigurationUtils.java:93]: getRuntimeVersion -> parseJavaRuntimeVersion

File: azure-toolkit-libs/azure-toolkit-springcloud-lib/src/main/java/com/microsoft/azure/toolkit/lib/springcloud/ConfigurationUtils.java [hunk 14]
```diff
@@ -0,0 +1,170 @@
...
+        if (matcher.matches()) {
+            return StringUtils.equals(matcher.group(4), "8") ? JavaRuntimeVersion.JAVA_8 : JavaRuntimeVersion.JAVA_11;
+        } else {
+            Log.warn(String.format("%s is not a valid runtime version, supported values are Java 8 and Java 11," +
+                " using Java 8 in this deployment.", fixedRuntimeVersion));
+            return DEFAULT_RUNTIME_VERSION;
+        }
+    }
+
+    public static AppResourceProperties getOrNewProperties(
```
Comment (andxu) [inline azure-toolkit-libs/azure-toolkit-springcloud-lib/src/main/java/com/microsoft/azure/toolkit/lib/springcloud/ConfigurationUtils.java:108]: getOrDefault

File: azure-toolkit-libs/azure-toolkit-springcloud-lib/src/main/java/com/microsoft/azure/toolkit/lib/springcloud/ConfigurationUtils.java [hunk 15]
```diff
@@ -0,0 +1,170 @@
...
+    ) {
+        AppResourceProperties properties = resource.properties();
+        if (Objects.isNull(properties)) {
+            properties = new AppResourceProperties();
+            resource.withProperties(properties);
+        }
+        return properties;
+    }
+
+    public static DeploymentResourceProperties getOrNewProperties(
```
Comment (andxu) [inline azure-toolkit-libs/azure-toolkit-springcloud-lib/src/main/java/com/microsoft/azure/toolkit/lib/springcloud/ConfigurationUtils.java:120]: getOrDefaultDeployResourceProerties

File: azure-toolkit-libs/azure-toolkit-springcloud-lib/src/main/java/com/microsoft/azure/toolkit/lib/springcloud/SpringCloudApp.java [hunk 16]
```diff
@@ -0,0 +1,180 @@
...
+import com.microsoft.azure.toolkit.lib.common.IEntityManager;
+import com.microsoft.azure.toolkit.lib.springcloud.model.SpringCloudAppEntity;
+import com.microsoft.azure.toolkit.lib.springcloud.model.SpringCloudDeploymentEntity;
+import com.microsoft.azure.toolkit.lib.springcloud.service.SpringCloudAppService;
+import lombok.Getter;
+
+import javax.annotation.Nonnull;
+import java.util.Objects;
+
+interface ISpringCloudAppModifier {
```
Comment (andxu) [inline azure-toolkit-libs/azure-toolkit-springcloud-lib/src/main/java/com/microsoft/azure/toolkit/lib/springcloud/SpringCloudApp.java:23]: ISpringCloudAppModifier -> ISpringCloudAppUpdate

File: azure-toolkit-libs/azure-toolkit-springcloud-lib/src/main/java/com/microsoft/azure/toolkit/lib/springcloud/SpringCloudCluster.java [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+
+    public SpringCloudClusterEntity entity() {
+        return Objects.nonNull(this.remote) ? this.remote : this.local;
+    }
+
+    public SpringCloudApp app(SpringCloudAppEntity app) {
+        return new SpringCloudApp(app, this);
+    }
+
+    public SpringCloudCluster reload() {
```
Comment (andxu) [inline azure-toolkit-libs/azure-toolkit-springcloud-lib/src/main/java/com/microsoft/azure/toolkit/lib/springcloud/SpringCloudCluster.java:43]: reload -> refresh?

File: azure-toolkit-libs/azure-toolkit-springcloud-lib/src/main/java/com/microsoft/azure/toolkit/lib/springcloud/SpringCloudDeployment.java [hunk 18]
```diff
@@ -0,0 +1,183 @@
...
+        return new Creator(this);
+    }
+
+    public Updater update() {
+        return new Updater(this);
+    }
+
+    @Override
+    public SpringCloudDeployment configEnvironmentVariables(Map<String, String> env) {
+        return this.update().configEnvironmentVariables(env).commit();
```
Comment (andxu) [inline azure-toolkit-libs/azure-toolkit-springcloud-lib/src/main/java/com/microsoft/azure/toolkit/lib/springcloud/SpringCloudDeployment.java:90]: can we hold the changes in batch and commit in final step?

File: azure-toolkit-libs/azure-toolkit-springcloud-lib/src/main/java/com/microsoft/azure/toolkit/lib/springcloud/SpringCloudDeployment.java [hunk 19]
```diff
@@ -0,0 +1,183 @@
...
+        public Modifier configAppArtifactPath(String path) {
+            final UserSourceInfo source = ConfigurationUtils.getOrNewSource(this.resource, this.deployment);
+            source.withRelativePath(path);
+            return this;
+        }
+
+        public abstract SpringCloudDeployment commit();
+    }
+
+    public static class Updater extends Modifier {
```
Comment (andxu) [inline azure-toolkit-libs/azure-toolkit-springcloud-lib/src/main/java/com/microsoft/azure/toolkit/lib/springcloud/SpringCloudDeployment.java:162]: what the differences between Updater  and Modifier ?

File: azure-toolkit-libs/azure-toolkit-springcloud-lib/src/main/java/com/microsoft/azure/toolkit/lib/common/IAzureEntity.java [hunk 20]
```diff
@@ -0,0 +1,26 @@
...
+ * THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+ * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package com.microsoft.azure.toolkit.lib.common;
+
+public interface IAzureEntity {
+}
```
Comment (shenqianjin) [inline azure-toolkit-libs/azure-toolkit-springcloud-lib/src/main/java/com/microsoft/azure/toolkit/lib/common/IAzureEntity.java:26]: is it necessary ?

Comment (shenqianjin) [reply azure-toolkit-libs/azure-toolkit-springcloud-lib/src/main/java/com/microsoft/azure/toolkit/lib/common/IAzureEntity.java:26]: If it is, I prefer to rename is as BaseEntity. abstract class.

File: azure-toolkit-libs/azure-toolkit-springcloud-lib/src/main/java/com/microsoft/azure/toolkit/lib/springcloud/SpringCloudApp.java [hunk 21]
```diff
@@ -0,0 +1,172 @@
...
+import java.util.Objects;
+
+interface ISpringCloudAppUpdater {
+
+    ISpringCloudAppUpdater activate(String deploymentName);
+
+    ISpringCloudAppUpdater setPublic(boolean isPublic);
+
+    ISpringCloudAppUpdater enablePersistentDisk(boolean enable);
+}
```
Comment (shenqianjin) [inline azure-toolkit-libs/azure-toolkit-springcloud-lib/src/main/java/com/microsoft/azure/toolkit/lib/springcloud/SpringCloudApp.java:30]: I feel that "updator" sounds strange.

File: azure-toolkit-libs/azure-toolkit-springcloud-lib/src/main/java/com/microsoft/azure/toolkit/lib/springcloud/model/SpringCloudAppConfig.java [hunk 22]
```diff
@@ -0,0 +1,29 @@
...
+    private final String appName;
+    private final String resourceGroup;
+    private final Boolean isPublic;
+    private final String runtimeVersion;
+    private final String activeDeploymentName;
+    private final SpringCloudDeploymentConfig deployment;
+
+    public Boolean isPublic() {
+        return BooleanUtils.isTrue(isPublic);
+    }
```
Comment (shenqianjin) [inline azure-toolkit-libs/azure-toolkit-springcloud-lib/src/main/java/com/microsoft/azure/toolkit/lib/springcloud/model/SpringCloudAppConfig.java:28]: why we need this method ?

**Comment by wangmingliang-ms** [comment]
this PR is only for preview

**Pull request closed without merging**